### ISSUE TITLE
How about the performance without pretrained weight?

### ISSUE THREAD

**Issue by twmht** [shamangary/SSR-Net#17]
How about the performance without the pretrained weight (that is, training from scratch)?

**Comment by shamangary** [comment]
Hello. I don't have the exact number now, but I remember it should be a little worse. We use the pretraining as indicated in our paper and the previous papers. Cheers.

**Comment by twmht** [comment]
@shamangary 

還有其他問題我直接用中文問好了，paper 裡面的 Section 3.5 的 Equation，age 的值算出來是有可能大於 V ，甚至是負的？ delta 的值吐出來可能會是 -1，這會有除0的問題(https://github.com/shamangary/SSR-Net/blob/master/training_and_testing/SSRNET_model.py#L154)？我看 code 的話你沒對這些問題做過處理。

我現在用 pytorch 複製整個 work，不用 pretrained weght，發現如果不去 clamp age 的話 (限制輸出最小是1，最大是V)，loss 容易 nan。

很好奇你有沒有碰過這些問題

**Comment by shamangary** [comment]
哈囉，delta可能是-1的部分我有考慮過，由於我之前有在該值前面乘上一個大約是0.99的值，所以分母數學上不會爆掉，但是後來我心血來潮把它拿掉，他還是沒有爆掉，所以我後來就省略了，不過你說得對，應該要保證分母不為0。

至於age值算出來可能會大於V這點我沒有遇到過，基本上這控制在p和s＿bar上比較多，而且像我上述提到的，由於delta是由tanh出來的，應該介於[-1,1]之間，你也可以用乘上一個值去調整他可能變動的範圍，理論上來說應該會比較穩定。

另外也要注意前處理的部分，臉部要特徵點align好會比較好，像是我在demo就省略這步驟可能會有很多問題，不過在train和test理論上是都要crop+align好的。

**Comment by twmht** [comment]
感謝回覆，age 是不是應該算出來（在乘上V之前）要介於 [0-1] 的區間會比較好？這樣乘上 V( V=100) 的話，age 才會被 scale 到 [0-100]?

p 如果是用 relu 的話，age 的 upper bound 不好算，如果是從分類的角度來看的話，是不是應該要接 softmax 會比較好？我用 softmax 取代 relu，看起來穩定很多。

**Comment by shamangary** [comment]
其實就是，因為network會自己調整，就算不限制[0,1]他也會符合訓練結果，不過你當然也可以先rescale，而dynamic scaling/indexing就保障了你亂設也沒差（咦？），會有一定自動調整的空間，畢竟能夠將人工減少到最多就是我們想要的。

softmax我也試過，只是因為整體來說並沒有太大差別，當然對公式也沒有影響，如我上述，如果p是介於[0,inf]或是[0,1]都會經過s_bar和i_bar的調整，我是覺得只要抓到精神小地方當然可以有各種變化。希望有幫助。 :)

**Comment by twmht** [comment]
@shamangary 

有一點我很好奇的是，你有測試過 ssrnet + 一般的 regression 嗎？ 因為我看 mobilenet 是用一般的 regression，這樣子比起來好像不太公平

不過原因也可能是因為 stepwise regression 必須要配 ssrnet

我的意思是說，從論文上似乎看不出來到底是 ssrnet 這種 two stream 的架構對抵抗 overfitting 有幫助，還是 stepwise regression 有幫助？

**Comment by shamangary** [comment]
ssrnet稱呼的是後面的stepwise regression+dynamic scaling/indexing，所以不存在所謂的"ssrnet + 一般的 regression"

two stream架構是來自於我上一篇DeepCD的啟發，用one stream也可以跑ssrnet

我做過分析如果把stepwise中的dynamic的part完全關掉mae會直接跳到9，該點有在paper中提到

**Comment by twmht** [comment]
@shamangary 

one stream 如果要跑 ssr net 的話，fusion block 也是有兩條 branch，還是只有一條？

那你有試過 mobilenet + ssr 嗎？

**Comment by trikim** [comment]
@twmht 你好，你把模型转成pytorch后最终MAE能达到什么程度？
